keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,706 @@
+  - if name == "foo", then field X must be nil, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/apiserver/validate_apiserver.go#L177)
+  - validate new label selector format: metav1.LabelSelector, this matches our deployment API
+  - validate old label selector format: map[string]string, this matches our node selectors for pods
+  - len(A)>0 xor len(B)>0
+  - quantity validation to match quota and usage specifications, [buried inside of here (most of it is not generally applicable)](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/clusterresourcequota/validation/validation.go#L44)
+  - if old value is X, don't allow changing the value, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/features/validate_features.go#L82)
+  - shorthand to match our NameIsDNSSubdomain and the like would be nice.
+  - if X exists in list1, then X must not existing in list2, both list1 and list2 are in the manifest, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/securitycontextconstraints/validation/validation.go#L139)
+- Use case: [Tekton pipeline validation](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipeline_validation.go)
+  - Referential integrity checks
+  - Custom formatted validation error messages
+  - "Either list X or list Y must be non-empty"
+  - "There exists a X in one list/map for each Y in another map/list"
+- (PRs to add additional user stories to this list welcome!)
+
+### Notes/Constraints/Caveats (Optional)
+
+While we believe the expressiveness of CEL is pretty complete for our purposes, it is non-turing
+complete, and lacks support recursive data types (OpenAPIv3 & CRD structural schemas are not
+possible to validate with CEL).
+
+### Risks and Mitigations
+
+#### Accidental misuse
+
+Break the control plane by consuming excessive CPU and/or memory the api-server.
+
+Mitigation: CEL is specifically designed to constrain the running time of expressions and to limit
+the memory utilization. We will run a series of performance benchmarks with CEL programs designed
+utilize a range of CPU and memory resources and document the results of the benchmarks before
+promoting this feature to GA.
+
+#### Malicious use
+
+Breaking out of the sandbox to run untrusted code in the apiserver or exfiltrate data.
+
+Mitigation: CEL is designed to sandbox code execution. Also, because CRD createion is a privileged
+operation, it should be safe to integrate.
+
+Additional limits we can put in place, as needed, include:
+- A max execution time limit to but could bound running time of CEL programs. This would require
```

Comment:
which means a CEL program may timeout. Similar to webhook, there may need to be a policy on whether to deny or allow when the validation fails.

Reply:
I don't think we'd ever allow if a validation rule timed out.

Reply:
Good point.

Reply:
Which means [failure policy](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy) will not be implemented here

Reply:
CEL evaluations are multiple orders of magnitude faster than typical webhook invocations... I suspect we'd bound the CEL expression complexity and handle timeouts the same way we would if etcd caused a write operation to timeout by returning a generic server timeout error

Reply:
It make perfect sense to treat runtime CEL failures as an internal error, just like the write timeout case. I think it would still be helpful to mention that in the KEP.

Reply:
Good news. The complexity estimation is already there. https://github.com/jinmmin/cel-go/blob/a661c99f8e27676c70fc00f4f328476ca4dcdb7f/cel/program.go#L265

Reply:
+1 to using complexity estimations. I think I also agree that a server timeout error is appropriate.  I'll make some notes in the KEP.